psqlgraph/node.py

```diff
@@ -204,6 +215,43 @@ def edge_predicate(e):
                     queue.append((n, depth + 1))
                     marked.add(n.node_id)
 
+    def _dfs(self, edge_predicate=None, edge_pointer="in", visited=None):
+        """
+        Perform a BFS, with `self` being the root node
+
+        :param edge_predicate: a predicate performed on an `edge` object in
+            order to decided whether to walk that edge or not
+        :type edge_predicate: func
+        :param edge_pointer: possible values `in`, `out`
+                            `in`: use node.edges_in, default behavior
+                            `out`: use edges_out
+        :type edge_pointer: str
+        :param visited: visited node ids
+        :type visited: set
+
+        :return: generator
+        """
+
+        if not callable(edge_predicate):
+            edge_predicate = lambda e: True
+
+        visited = visited or set()
+        visited.add(self.node_id)
+
+        yield self
+
+        edges = self.edges_out if edge_pointer == "out" else self.edges_in
+        for edge in edges:
+            if not edge_predicate(edge):
+                continue
+
+            n = edge.dst if edge_pointer == "out" else edge.src
+
+            if n.node_id not in visited:
+                res = n._dfs(edge_predicate, edge_pointer, visited)
```

Comment:
I'd prefer if this were not recursive. Since this library is intended for general use, we shouldn't assume that the max depth of the graph is less than the max recursion limit of the python interpreter.

In iterative approach to dfs is not as elegant, but it'll be much safer.